Gender NeoPronoun Transform

jjangsangy opened the pull request:
Neopronouns are a category of new (neo) pronouns that are increasingly used in place of “she,” “he,” or “they” when referring to a person. Some examples include: xe/xem/xyr, ze/hir/hirs, and ey/em/eir. Neopronouns can be used by anyone, though most often they are used by transgender, non-binary, and/or gender nonconforming people.

This transformation performs **_grammatically correct substitution_** of the gendered pronouns he/she in a given sentence with their neopronoun counterparts. The reason why a simple look-up-table approach might not work as depending on the context, the _case_ can belong to the _Nominative/Subject Pronouns,	Accusative/Object Pronouns,	Pronominal Possessive/Possessive Adjectives,	Predicative Possessive/Possessive Pronouns	or Reflexive Pronouns_ categories.

@vinayprabhu @rteehas @nick11roberts

jjangsangy commented:
@vinayprabhu

sebastianGehrmann commented on transformations/gender_neopronouns/README.md:
```diff
@@ -0,0 +1,83 @@
+# Neopronoun substitution 🏳️‍⚧️
+
+Authors:
```
Would you mind adding your affiliation as well?

sebastianGehrmann commented on transformations/gender_neopronouns/README.md:
```diff
@@ -0,0 +1,83 @@
+Answer: _Neopronouns are a category of new (neo) pronouns that are increasingly used in place of “she,” “he,” or “they” when
+referring to a person. Some examples include: xe/xem/xyr, ze/hir/hirs, and ey/em/eir. Neopronouns can be used by
+anyone, though most often they are used by transgender, non-binary, and/or gender nonconforming people._
+Source: https://intercultural.uncg.edu/wp-content/uploads/Neopronouns-Explained-UNCG-Intercultural-Engagement.pdf
```
Can you format this as a [link](https://intercultural.uncg.edu/wp-content/uploads/Neopronouns-Explained-UNCG-Intercultural-Engagement.pdf)?

sebastianGehrmann commented on transformations/gender_neopronouns/README.md:
```diff
@@ -0,0 +1,83 @@
+anyone, though most often they are used by transgender, non-binary, and/or gender nonconforming people._
+Source: https://intercultural.uncg.edu/wp-content/uploads/Neopronouns-Explained-UNCG-Intercultural-Engagement.pdf
+
+In the image below, we can find the temporal variations in neopronoun usage over the past few years from  the **_Gender Census 2021: Worldwide Report_**.
```
Can you add a sentence here with an interpretation of the image?

sebastianGehrmann commented on transformations/gender_neopronouns/README.md:
```diff
@@ -0,0 +1,83 @@
+This transformation seeks to render the training data used in NLP pipelines more neopronoun aware so as to help tackle the downstream risks of [trans erasure](https://allthingslinguistic.com/post/118373603278/i-am-at-the-end-of-my-first-semester-of).
+
+Benchmark Results:
+* Seq2Seq Generation task ran on IMDB dataset/Roberta model with a 1% sample size. The original accuracy was 96% and the perterbed accuracy was 95%
```
nit: perterbed-> perturbed

sebastianGehrmann replied:
Also, I am unsure what you mean by Seq2Seq model on IMDB with Roberta. This is a classification task and a masked language model, both of which are not seq2seq.

sebastianGehrmann commented on transformations/gender_neopronouns/README.md:
```diff
@@ -0,0 +1,83 @@
+<img title=" The neopronoun preferences over times" src="https://gendercensus.files.wordpress.com/2021/03/gc2021-pronouns-over-time-minus-he-she-they.png" width="600" height="354" />
+
+## What type of a transformation is this?
+This transformation performs **_grammatically correct substitution_** of the gendered pronouns he/she in a given sentence with their neopronoun counterparts. The reason why a simple look-up-table approach might not work as depending on the context, the _case_ can belong to the _Nominative/Subject Pronouns,	Accusative/Object Pronouns,	Pronominal Possessive/Possessive Adjectives,	Predicative Posses ...
```
#benderrule - Can you specifically state that this is for English?

sebastianGehrmann commented on transformations/gender_neopronouns/test.json:
```diff
@@ -0,0 +1,41 @@
+          "sentence": "I have to drive myself to school."
+        },
+        "outputs": [{
+          "sentence": "Ey have to drive emself to school."
```
Can you add a test case with the third person singular s, like a transformation of "She walks to the mall" to "they"?

sebastianGehrmann commented on transformations/gender_neopronouns/transformation.py:
```diff
@@ -0,0 +1,85 @@
+                        else neopronoun.lower()
+                    )
+                    pieces.append(neopronoun)
+                except Exception:
```
Please specify which type of Exception.

sebastianGehrmann commented on transformations/gender_neopronouns/transformation.py:
```diff
@@ -0,0 +1,85 @@
+
+        super().__init__(seed=seed, verbose=verbose, max_outputs=max_outputs)
+
+    def generate(self, sentence: str) -> List[str]:
```
Please add a docstring

sebastianGehrmann commented on transformations/gender_neopronouns/transformation.py:
```diff
@@ -0,0 +1,85 @@
+                else:
+                    pron_type = "ACC"
+
+                try:
```
I may have missed it but where is the third person singular s handling here?

jjangsangy replied:
I added the section for third person singular handling below in the `elif` clause.

sebastianGehrmann reviewed (changes requested):
Fantastic PR, thank you for the contribution! I left a bunch of minor comments.

james-simon commented:
I'm here to leave a review. Thanks for the timely task, and thanks @sebastianGehrmann for taking a close look at the code! This looks like a high-quality task that should end up in the benchmark. I have only a few high-level points to address.
* Your README mentions that, since the same standard pronoun is often used in multiple cases, this task needs grammar-aware substitution. It'd help clarify and emphasize this fact if you gave an example sentence (e.g. "Her dog nuzzled her") that illustrates that ambiguity.
* In the same vein, you should add a few tests that show that your code actually does identify the right pronoun cases despite that ambiguity.
* Your README states that "this transformation performs grammatically correct substitution of the gendered pronouns he/she in a given sentence with their neopronoun counterparts," but, judging by your tests, your code swaps out first-person and they/... pronouns, too. Which is it meant to do? Also, just out of curiosity, do neopronoun users often replace the standard first-person pronouns?

Let me know if you have any questions!

jjangsangy commented:
@sebastianGehrmann I've gone and addressed the minor issues and documentation. I'm currently working on the issue with the first person singular as well as the wrong cases mentioned by @james-simon. Thanks for both of your helpful inputs!

asnota commented on transformations/gender_neopronouns/transformation.py:
```diff
@@ -0,0 +1,116 @@
+
+    tasks = [TaskType.TEXT_TO_TEXT_GENERATION]
+    languages = ["en"]
+
```
Please provide relevant keywords, using [this list](https://github.com/GEM-benchmark/NL-Augmenter/blob/main/docs/keywords.md)

jjangsangy commented:
@james-simon thanks for the helpful comments.

* I've addressed the issue with the grammar-aware case of "Her dog nuzzled her". It now produces the correct substitution to be "Their dog nuzzled them".
* I also added 2 more additional test cases to cover this ambiguous case.

> Also, just out of curiosity, do neopronoun users often replace the standard first-person pronouns?
* First person pronouns should not be swapped, this was an oversight on my part. This is now reflected in the code and should pass through and not be replaced by a neopronoun.

james-simon commented:
Thanks for the changes! For completeness, you should probably also add tests showing that 2nd-person and 1st-person-plural pronouns pass through untransformed. Trusting you'll do that (or explain why it doesn't matter), I'll now accept the task.

jjangsangy commented:
Gotcha! I've implemented filtering of second person pronouns as well and wrote the test case showing you/your pass through.

sebastianGehrmann commented:
Thanks for all the changes! Looking great to me know. If the tests pass, I am happy with it being merged. Thank you for your contribution!

jjangsangy commented:
Tests have passed but it looks like we need 2 reviews to successfully complete the merge.
